src/software/simulated_tests/simulated_test_fixture.py

```diff
@@ -94,133 +100,153 @@ def time_provider(self):
         with self.timestamp_mutex:
             return self.timestamp
 
-    def run_test(
+    def excepthook(self, args):
+        """This function is _critical_ for show_thunderscope to work.
+        If the test Thread will raises an exception we won't be able to close
+        the window from the main thread.
+
+        :param args: The args passed in from the hook
+
+        """
+
+        __stopper(delay=PAUSE_AFTER_FAIL_DELAY_S)
+        self.last_exception = args.exc_value
+        raise self.last_exception
+
+    def __stopper(self, delay=PROCESS_BUFFER_DELAY_S):
+        """Stop running the test
+
+        :param delay: How long to wait before closing everything, defaults
+                      to PROCESS_BUFFER_DELAY_S to minimize buffer warnings
+
+        """
+        time.sleep(delay)
+
+        if self.thunderscope:
+            self.thunderscope.close()
+
+    def runner(
         self,
         always_validation_sequence_set=[[]],
         eventually_validation_sequence_set=[[]],
         test_timeout_s=3,
         tick_duration_s=0.0166,  # Default to 60hz
     ):
-        """Run a test
-
-        :param always_validation_sequence_set: Validation functions that should
-                                hold on every tick
-        :param eventually_validation_sequence_set: Validation that should
-                                eventually be true, before the test ends
-        :param test_timeout_s: The timeout for the test, if any eventually_validations
-                                remain after the timeout, the test fails.
-        :param tick_duration_s: The simulation step duration
-
+        """Step simulation, full_system and run validation
         """
 
-        def __stopper(delay=PROCESS_BUFFER_DELAY_S):
-            """Stop running the test
+        time_elapsed_s = 0
+
+        while time_elapsed_s < test_timeout_s:
 
-            :param delay: How long to wait before closing everything, defaults
-                          to PROCESS_BUFFER_DELAY_S to minimize buffer warnings
+            # Update the timestamp logged by the ProtoLogger
+            with self.timestamp_mutex:
+                ssl_wrapper = self.ssl_wrapper_buffer.get(block=False)
+                self.timestamp = ssl_wrapper.detection.t_capture
 
-            """
-            time.sleep(delay)
+            tick = SimulatorTick(milliseconds=tick_duration_s * MILLISECONDS_PER_SECOND)
+            self.simulator_proto_unix_io.send_proto(SimulatorTick, tick)
+            time_elapsed_s += tick_duration_s
 
             if self.thunderscope:
-                self.thunderscope.close()
+                time.sleep(tick_duration_s)
```

Comment:
what's this for?

Reply:
this was in the old implementation of the test runner, so I kept it the same. I can't run thunderscope right now as I'm running into an error, but it looks like maybe thunderscope can only receive validation protos at 60Hz and not faster so this artificial delay is there